adam_core/utils/indexable.py

```diff
@@ -1,174 +1,446 @@
+import logging
 from collections import OrderedDict
-from copy import deepcopy
-from typing import List, Union
+from copy import copy, deepcopy
+from typing import List, Optional, Union
 
 import numpy as np
+import numpy.typing as npt
 import pandas as pd
 from astropy.time import Time
 
 __all__ = ["Indexable", "concatenate"]
 
+logger = logging.getLogger(__name__)
+
+SLICEABLE_DATA_STRUCTURES = (np.ndarray, np.ma.masked_array, Time)
 UNSLICEABLE_DATA_STRUCTURES = (str, int, float, dict, bool, set, OrderedDict)
 
 
 class Indexable:
     """
     Class that enables indexing and slicing of itself and its members.
-    If an Indexable subclass has members that are `~numpy.ndarray`s,
-    `~numpy.ma.core.MaskedArray`s, lists, or `~astropy.time.core.Time`s
-    then these members are appropriately sliced and indexed along their first axis.
-    Any members that are dicts, OrderedDicts, floats, integers or strings are
-    not indexed and left unchanged.
+    If an Indexable subclass has members that are `~numpy.ndarray`s, `~numpy.ma.core.MaskedArray`s,
+    lists, or `~astropy.time.core.Time`s then these members are appropriately sliced and
+    indexed along their first axis. Any members that are dicts, OrderedDicts, floats,
+    integers or strings are not indexed and left unchanged.
+
+    Indexable maintains two array indexes:
+        - class: the externally facing index (ie. the index into the class itself)
+        - members: the index of the members of the class (ie. the index into
+            the data structures carried by the class)
+
+
+    Class Index:
+    [0, 1, 2]
+    The class index is used to determine the length (size) of the class. In this example,
+    the len(instance) would be 3.
+
+    Class To Members Mapping:
+    [0, 0, 0, 1, 1, 1, 2, 2, 2]
+    If, as above, the class to members mapping is sorted (monotonically increasing), then
+    it is converted to an array of slices which allows for faster querying.
+    [slice(0, 2), slice(2, 5), slice(5, 8)]
+
+    Members Index:
+    [0, 1, 2, 3, 4, 5, 6, 7, 8]
```

Comment:
I don't really know what this means.

Is class something like the first dimension - the number of columns? A more concrete example would help.

Calling it "class" for the "externally facing index" makes me think that a better name might be "container". "Class" makes me think of class-level (as opposed to instance-level) behavior, which this definitely has nothing to do with.

Reply:
Added a clearer example in the new docstring.